dGame/dCinema/Recorder.cpp

```diff
@@ -0,0 +1,1195 @@
+		controllableComponent->SetAngularVelocity(angularVelocity);
+		controllableComponent->SetIsOnGround(onGround);
+		controllableComponent->SetDirtyVelocity(dirtyVelocity);
+		controllableComponent->SetDirtyAngularVelocity(dirtyAngularVelocity);
+	}
+
+	Game::entityManager->SerializeEntity(actor);
+}
+
+void MovementRecord::Serialize(tinyxml2::XMLDocument& document, tinyxml2::XMLElement* parent) {
+	auto* element = document.NewElement("MovementRecord");
+
+	element->SetAttribute("x", position.x);
+	element->SetAttribute("y", position.y);
+	element->SetAttribute("z", position.z);
+
+	element->SetAttribute("qx", rotation.x);
+	element->SetAttribute("qy", rotation.y);
+	element->SetAttribute("qz", rotation.z);
+	element->SetAttribute("qw", rotation.w);
+
+	element->SetAttribute("vx", velocity.x);
+	element->SetAttribute("vy", velocity.y);
+	element->SetAttribute("vz", velocity.z);
+
+	element->SetAttribute("avx", angularVelocity.x);
+	element->SetAttribute("avy", angularVelocity.y);
+	element->SetAttribute("avz", angularVelocity.z);
+
+	element->SetAttribute("g", onGround);
+
+	element->SetAttribute("dv", dirtyVelocity);
+
+	element->SetAttribute("dav", dirtyAngularVelocity);
+
+	element->SetAttribute("t", m_Delay);
+
+	parent->InsertEndChild(element);
+}
+
+void MovementRecord::Deserialize(tinyxml2::XMLElement* element) {
+	position.x = element->FloatAttribute("x");
+	position.y = element->FloatAttribute("y");
+	position.z = element->FloatAttribute("z");
+
+	rotation.x = element->FloatAttribute("qx");
+	rotation.y = element->FloatAttribute("qy");
+	rotation.z = element->FloatAttribute("qz");
+	rotation.w = element->FloatAttribute("qw");
+
+	velocity.x = element->FloatAttribute("vx");
+	velocity.y = element->FloatAttribute("vy");
+	velocity.z = element->FloatAttribute("vz");
+
+	angularVelocity.x = element->FloatAttribute("avx");
+	angularVelocity.y = element->FloatAttribute("avy");
+	angularVelocity.z = element->FloatAttribute("avz");
+
+	onGround = element->BoolAttribute("g");
+
+	dirtyVelocity = element->BoolAttribute("dv");
+
+	dirtyAngularVelocity = element->BoolAttribute("dav");
+
+	m_Delay = element->DoubleAttribute("t");
+}
+
+SpeakRecord::SpeakRecord(const std::string& text) {
+	this->text = text;
+}
+
+void SpeakRecord::Act(Entity* actor) {
+	GameMessages::SendNotifyClientZoneObject(
+		actor->GetObjectID(), u"sendToclient_bubble", 0, 0, actor->GetObjectID(), text, UNASSIGNED_SYSTEM_ADDRESS);
+
+	Game::entityManager->SerializeEntity(actor);
+}
+
+void SpeakRecord::Serialize(tinyxml2::XMLDocument& document, tinyxml2::XMLElement* parent) {
+	auto* element = document.NewElement("SpeakRecord");
+
+	element->SetAttribute("text", text.c_str());
+
+	element->SetAttribute("t", m_Delay);
+
+	parent->InsertEndChild(element);
+}
+
+void SpeakRecord::Deserialize(tinyxml2::XMLElement* element) {
+	text = element->Attribute("text");
+
+	m_Delay = element->DoubleAttribute("t");
+}
+
+AnimationRecord::AnimationRecord(const std::string& animation) {
```

Comment:
May be worth splitting this into smaller files, this file is doing quite a lot of work right now with how large it is